testing/e2e/agent_install_test.go

```diff
@@ -0,0 +1,348 @@
+	suite.Require().NoError(err)
+}
+
+// extractDarwin treats the passed Reader as a tar.gz stream and unarchives it to a temp dir
+// fleet-server binary in archive is replaced by a locally compiled version
+// Additionally the MacOS specific elastic-agent symlink will be recreated with a hard link.
+func (suite *AgentInstallSuite) extractDarwin(r io.Reader) {
+	suite.T().Helper()
+	var fleetPath, agentSrc, agentDst, agentLink string
+	// Extract tar.gz stream
+	stream, err := gzip.NewReader(r)
+	suite.Require().NoError(err)
+	tarReader := tar.NewReader(stream)
+	for header, err := tarReader.Next(); err == nil; header, err = tarReader.Next() {
+		if header.FileInfo().IsDir() {
+			err := os.Mkdir(filepath.Join(suite.downloadPath, header.Name), 0755)
+			suite.Require().NoError(err)
+		} else {
+			if !header.FileInfo().Mode().IsRegular() && header.FileInfo().Name() == suite.agentName {
+				// MacOS archives have a symlink for the elastic-agent binary
+				agentDst = filepath.Join(suite.downloadPath, header.Name)
+				agentLink = header.Linkname
+				continue
+			} else if !header.FileInfo().Mode().IsRegular() {
+				suite.T().Logf("unable to extract %s", header.Name)
+				continue
+			}
+			dst, err := os.Create(filepath.Join(suite.downloadPath, header.Name))
+			suite.Require().NoError(err)
+			err = dst.Chmod(header.FileInfo().Mode())
+			suite.Require().NoError(err)
+			_, err = io.Copy(dst, tarReader)
+			dst.Close() // might be a dirty close
+			suite.Require().NoError(err)
+
+			if strings.HasSuffix(header.Name, binaryName) {
+				fleetPath = filepath.Join(suite.downloadPath, header.Name)
+			}
+			if strings.HasSuffix(header.Name, agentLink) {
+				agentSrc = filepath.Join(suite.downloadPath, header.Name)
+				suite.T().Logf("source is %s", agentSrc)
+			}
+		}
+	}
+	// Copy fleet-server binary to un archived package
+	suite.Require().NotEmpty(fleetPath, "no fleet-server component detected")
+	err = os.Remove(fleetPath)
+	suite.Require().NoError(err)
+	err = os.Link(suite.binaryPath, fleetPath)
+	suite.Require().NoError(err)
+
+	// link elastic-agent to the actual binary
+	suite.Require().NotEmpty(agentSrc, "agent link src undetected")
+	suite.Require().NotEmpty(agentDst, "agent link dst undetected")
+	err = os.Link(agentSrc, agentDst)
+	suite.Require().NoError(err)
+}
+
+// extractLinux treats the passed Reader as a tar.gz stream and unarchives it to a temp dir
+// Additionally the MacOS specific elastic-agent symlink will be recreated with a hard link.
+func (suite *AgentInstallSuite) extractLinux(r io.Reader) {
+	suite.T().Helper()
+	fleetPath := ""
+	// Extract tar.gz stream
+	stream, err := gzip.NewReader(r)
+	suite.Require().NoError(err)
+	tarReader := tar.NewReader(stream)
+	for header, err := tarReader.Next(); err == nil; header, err = tarReader.Next() {
+		if header.FileInfo().IsDir() {
+			err := os.Mkdir(filepath.Join(suite.downloadPath, header.Name), 0755)
+			suite.Require().NoError(err)
+		} else {
```

Comment:
Nit. Save a nesting level here.
```suggestion
			suite.Require().NoError(err)
			continue
		}
```